Support build against sle12sp5

nick-wang opened the pull request:
Support build with make_build undefined and enable GNU c99.

Backport clean func sd_bus_message_unrefp and sd_bus_flush_close_unrefp

nick-wang committed:
Support build with make_build undefined and enable GNU c99.

Error messages:
src/main.c:94:2: error: 'for' loop initial declarations are only allowed in C99 mode
  for (size_t i = 0; i < n_pids; i++) {
  ^
src/main.c:94:2: note: use option -std=c99 or -std=gnu99 to compile your code

nick-wang committed:
Backport clean func sd_bus_message_unrefp and sd_bus_flush_close_unrefp

Error messages:
src/main.c: In function 'migrate':
src/main.c:31:2: error: cleanup argument not a function
  _cleanup_(sd_bus_message_unrefp) sd_bus_message *m = NULL;
  ^
src/main.c: In function 'main':
src/main.c:234:2: error: cleanup argument not a function
  _cleanup_(sd_bus_flush_close_unrefp) sd_bus *bus = NULL;
  ^

Werkov commented on rpm/sapwmp.spec:
```diff
@@ -61,7 +61,13 @@ Configuration and utilities for collecting SAP processes under control group to
 %build
 ./autogen.sh
 %configure
+# make_build not defined in SLE12, so need conditions to check
```
The [wiki](https://en.opensuse.org/openSUSE:Specfile_guidelines#Parallel_make) says that %make_build should be available in >= SLE12 SP3. We really don't want to go back that much into the past.
If the wiki's wrong and even SLE12 SP5 needs this, I'm fine with the change.

nick-wang replied:
In reality, this macro is not existed in SLE12SP5... rpm-4.11.2-16.21.1.x86_64

```
# rpmbuild --showrc |grep make
      echo "Running automake --gnu ${am_opt} ..."; 
      %{__automake} --add-missing --gnu ${am_opt}; 
-14: __automake	automake
-14: __make	/usr/bin/make
	    yes "" | make oldconfig 
	    make dep 
-14: make_install	make install DESTDIR=%{?buildroot}
-14: makeinstall	make DESTDIR=%{?buildroot:%{buildroot}} install
-14: perl_make_install	make DESTDIR=$RPM_BUILD_ROOT install_vendor
```

Werkov replied:
Ack.

Werkov commented on src/main.c:
```diff
@@ -19,6 +19,18 @@
 
 #define _cleanup_(x) __attribute__((cleanup(x)))
 
+#ifndef _SD_DEFINE_POINTER_CLEANUP_FUNC
```
The macro comes from libsystemd's `_sd-common.h` private header.
I'd add a comment with that path: a) gives credit, b) provides hints where to look at if systemd changes it in the future.

nick-wang replied:
Got it. Added in the new commit.

Werkov reviewed (changes requested):
Thanks. Just nitpicks.

nick-wang committed:
Add a comment for the reason/source of define _SD_DEFINE_POINTER_CLEANUP_FUNC.

nick-wang commented:
Still have two rpmlint issues need to be fixed to complete the SLE12SP5 build:
```
[   20s] sapwmp.x86_64: E: permissions-file-setuid-bit (Badness: 10000) /usr/lib/sapwmp/sapwmp-capture is packaged with setuid/setgid bits (04750)
[   20s] If the package is intended for inclusion in any SUSE product please open a bug
[   20s] report to request review of the program by the security team
[   20s] 
[   20s] sapwmp.x86_64: E: non-position-independent-executable (Badness: 10000) /usr/lib/sapwmp/sapwmp-capture
[   20s] This executable must be position independent.  Check that it is built with
[   20s] -fPIE/-fpie in compiler flags and -pie in linker flags.

```
`permissions-file-setuid-bit` need security team to involve.

nick-wang committed:
Enable position independent executable.

sapwmp.x86_64: E: non-position-independent-executable
               (Badness: 10000) /usr/lib/sapwmp/sapwmp-capture
This executable must be position independent.  Check that it is
built with -fPIE/-fpie in compiler flags and -pie in linker flags.

Werkov commented on configure.ac:
```diff
@@ -1,8 +1,12 @@
 AC_CONFIG_FILES(Makefile)
 
+AC_SUBST([CFLAGS], [-fPIE])
+AC_SUBST([LDFLAGS], [-pie])
```
Ack.

Werkov commented:
Thanks, LGTM (the rpmlint has to be resolved out of the package itself) -> merging.

nick-wang commented:
@Werkov Thanks for the review and information!
